File: bridge/src/platforms/slack.js
'use strict';

const { App } = require('@slack/bolt');
const { HttpsProxyAgent } = require('https-proxy-agent');
const { execute } = require('../executor');
const { sanitize } = require('../utils/sanitize');
const { chunk } = require('../utils/chunker');

const SLACK_MAX_LEN = 2900;

/**
 * Start the Slack platform adapter (Socket Mode).
 * @param {object} config - Slack configuration block from bridge.config.json.
 */
async function start(config) {
  if (!config || !config.enabled || !config.bot_token) {
    console.log('[slack] Disabled or no token — skipping.');
    return null;
  }

  const {
    bot_token,
    app_token,
    signing_secret,
    allowed_channel_ids = [],
    allowed_user_ids = [],
  } = config;

  if (!app_token) {
    console.error('[slack] app_token is required for Socket Mode — skipping.');
    return null;
  }

  const proxyUrl = process.env.HTTPS_PROXY || process.env.https_proxy;
  let agent;
  if (proxyUrl) {
    try {
      agent = new HttpsProxyAgent(proxyUrl);
      console.log(`[slack] Using proxy: ${proxyUrl}`);
    } catch (err) {
      console.error(`[slack] Invalid proxy URL in HTTPS_PROXY "${proxyUrl}": ${err.message}`);
      return null;
    }
  }

  const app = new App({
    token: bot_token,
    appToken: app_token,
    signingSecret: signing_secret,
    socketMode: true,
    agent,
  });

  /**
   * Check access and handle a prompt from a Slack event.
   */
  async function handleEvent({ event, say, client }) {
    const channelId = event.channel;
    const userId = event.user;
    const text = event.text || '';

    // Channel whitelist check
    if (allowed_channel_ids.length > 0 && !allowed_channel_ids.includes(channelId)) {
      return;
    }

    // User whitelist check
    if (allowed_user_ids.length > 0 && !allowed_user_ids.includes(userId)) {
      return;
    }

    // Strip bot mention from text
    const rawPrompt = text.replace(/<@[A-Z0-9]+>/g, '').trim();
    if (!rawPrompt) return;

    const { safe: prompt } = sanitize(rawPrompt);
    if (!prompt.trim()) return;

    // Send processing placeholder
    let processingTs;
    try {
      const sent = await say('⏳ 處理中...');
      processingTs = sent.ts;
    } catch (err) {
      console.error('[slack] Failed to send processing message:', err);
      return;
    }

    let result;
    try {
      result = await execute(prompt, null);
    } catch (err) {
      console.error('[slack] execute() threw:', err);
      try {
        await client.chat.update({ channel: channelId, ts: processingTs, text: '❌ 執行錯誤，請稍後再試。' });
      } catch (_) { /* best-effort */ }
      return;
    }

    const output = result.success
      ? result.output || '(no output)'
      : `Error: execution failed.\n${result.output}`;

    const chunks = chunk(output, SLACK_MAX_LEN);

    // Update the placeholder with the first chunk
    try {
      await client.chat.update({
        channel: channelId,
        ts: processingTs,
        text: chunks[0],
      });
    } catch (err) {
      console.error('[slack] Failed to update processing message:', err);
      // Fallback: send as new message so the result isn't lost
      try {
        await say(chunks[0]);
      } catch (e) {
        console.error('[slack] Failed to send fallback message:', e);
      }
    }

    // Post remaining chunks
    for (let i = 1; i < chunks.length; i++) {
      try {
        await say(chunks[i]);
      } catch (err) {
        console.error(`[slack] Failed to send chunk ${i}:`, err);
      }
    }
  }

  // Respond to app_mention events
  app.event('app_mention', async ({ event, say, client, ack }) => {
    if (typeof ack === 'function') await ack();
    await handleEvent({ event, say, client });
  });

  // Respond to DMs (message events in im channels)
  app.message(async ({ message, say, client, ack }) => {
    if (typeof ack === 'function') await ack();
    // Only handle DMs
    if (message.channel_type !== 'im') return;
    await handleEvent({ event: message, say, client });
  });

  app.error((err) => {
    console.error('[slack] App error:', err);
  });

  await app.start();
  console.log('[slack] App started (Socket Mode).');

  return app;
}

module.exports = { start };

File: bridge-go/slack.go
package main

import (
	"context"
	"fmt"
	"log/slog"
	"regexp"
	"slices"
	"strings"
	"sync"

	"github.com/slack-go/slack"
	"github.com/slack-go/slack/slackevents"
	"github.com/slack-go/slack/socketmode"
)

const slackChunkRunes = 2900

var slackMentionRe = regexp.MustCompile(`<@[A-Z0-9]+>`)

type Slack struct {
	cfg  SlackConfig
	exec ExecutorConfig
	log  *slog.Logger
	wg   sync.WaitGroup
}

func NewSlack(cfg SlackConfig, exec ExecutorConfig) *Slack {
	return &Slack{cfg: cfg, exec: exec, log: slog.With("platform", "slack")}
}

func (sl *Slack) Name() string { return "slack" }

func (sl *Slack) Run(ctx context.Context) error {
	if sl.cfg.AppToken == "" {
		return fmt.Errorf("app_token is required for Socket Mode")
	}

	api := slack.New(
		sl.cfg.BotToken,
		slack.OptionAppLevelToken(sl.cfg.AppToken),
	)

	client := socketmode.New(api,
		socketmode.OptionDebug(false),
	)

	// runErr receives the error from RunContext so we can surface it to main.
	runErr := make(chan error, 1)
	go func() {
		runErr <- client.RunContext(ctx)
	}()

	sl.log.Info("bot started")

	// Wait for in-flight handlers before returning.
	defer sl.wg.Wait()

	for {
		select {
		case err := <-runErr:
			// RunContext returned — either ctx cancelled (clean) or connection error.
			if ctx.Err() != nil {
				return nil
			}
			return fmt.Errorf("socket mode disconnected: %w", err)

		case <-ctx.Done():
			// Drain runErr to avoid goroutine leak.
			<-runErr
			return nil

		case evt, ok := <-client.Events:
			if !ok {
				return nil
			}
			switch evt.Type {
			case socketmode.EventTypeConnected:
				sl.log.Info("WebSocket connected")
			case socketmode.EventTypeDisconnect:
				sl.log.Info("WebSocket disconnected")
			case socketmode.EventTypeEventsAPI:
				client.Ack(*evt.Request)
				eventsAPIEvent, ok := evt.Data.(slackevents.EventsAPIEvent)
				if !ok {
					continue
				}
				sl.wg.Add(1)
				go func() {
					defer sl.wg.Done()
					defer func() {
						if r := recover(); r != nil {
							sl.log.Error("handler panic", "recover", r)
						}
					}()
					sl.handleEventsAPI(ctx, api, eventsAPIEvent)
				}()
			}
		}
	}
}

func (sl *Slack) handleEventsAPI(ctx context.Context, api *slack.Client, event slackevents.EventsAPIEvent) {
	if event.Type != slackevents.CallbackEvent {
		return
	}

	var channelID, userID, text string
	isDM := false

	switch ev := event.InnerEvent.Data.(type) {
	case *slackevents.AppMentionEvent:
		channelID = ev.Channel
		userID = ev.User
		text = ev.Text
	case *slackevents.MessageEvent:
		if ev.BotID != "" || ev.SubType != "" {
			return
		}
		if ev.ChannelType != "im" {
			return
		}
		isDM = true
		channelID = ev.Channel
		userID = ev.User
		text = ev.Text
	default:
		return
	}

	// Channel whitelist — skip for DMs (channelID is the DM channel, not a public channel ID)
	if !isDM && len(sl.cfg.AllowedChannelIDs) > 0 && !slices.Contains(sl.cfg.AllowedChannelIDs, channelID) {
		return
	}
	// User whitelist
	if len(sl.cfg.AllowedUserIDs) > 0 && !slices.Contains(sl.cfg.AllowedUserIDs, userID) {
		return
	}

	// Strip bot mention
	rawPrompt := slackMentionRe.ReplaceAllString(text, "")
	rawPrompt = strings.TrimSpace(rawPrompt)
	if rawPrompt == "" {
		return
	}

	prompt, truncated := sanitizePrompt(rawPrompt)
	if prompt == "" {
		return
	}

	// Notify user if prompt was truncated before processing.
	if truncated {
		if _, _, err := api.PostMessage(channelID, slack.MsgOptionText("⚠️ 您的訊息過長，已截斷至 10,000 字元。", false)); err != nil {
			sl.log.Error("send truncation notice failed", "err", err)
		}
	}

	// Send placeholder
	_, placeholderTS, err := api.PostMessage(channelID, slack.MsgOptionText("⏳ 處理中...", false))
	if err != nil {
		sl.log.Error("send placeholder failed", "err", err)
		return
	}

	result := Execute(ctx, sl.exec, prompt)
	sl.log.Info("execution complete", "duration_ms", result.DurationMs, "success", result.Success, "timed_out", result.TimedOut)
	output := result.FormatForDisplay()
	chunks := Chunk(output, slackChunkRunes)

	// Update placeholder with first chunk
	_, _, _, err = api.UpdateMessage(channelID, placeholderTS, slack.MsgOptionText(chunks[0], false))
	if err != nil {
		sl.log.Error("update placeholder failed", "err", err)
		// Fallback: post as new message
		if _, _, err2 := api.PostMessage(channelID, slack.MsgOptionText(chunks[0], false)); err2 != nil {
			sl.log.Error("fallback post failed", "err", err2)
			return
		}
	}

	// Send remaining chunks; stop and notify on failure
	for i, chunk := range chunks[1:] {
		if _, _, err := api.PostMessage(channelID, slack.MsgOptionText(chunk, false)); err != nil {
			sl.log.Error("send chunk failed", "chunk", i+1, "err", err)
			_, _, _ = api.PostMessage(channelID, slack.MsgOptionText("⚠️ 回覆傳送不完整，部分內容遺失。", false))
			return
		}
	}
}
